nixos/modules/services/networking/iwd.nix

```diff
@@ -4,16 +4,142 @@ with lib;
+
+  # Generates a name-value-pair with configurations for a SSID. For further information
+  # about the configuration options, please refer to the upstream docs:
+  #
+  # https://iwd.wiki.kernel.org/networkconfigurationsettings
+  predefinedNetworks = flip mapAttrs cfg.networks (name: cfg: {
+    extension = determineType cfg;
+
+    # Cannot use the SSID here: the hex-encode is done at build-time, not eval time
+    # and several special chars in a file name would break `writeText`.
+    file = pkgs.writeText "ssid" (generators.toINI {} ({
+      Settings.AutoConnect = cfg.autoConnect;
+      Settings.Hidden = cfg.hidden;
+    } // (optionalAttrs (cfg.psk != null) {
+      Security.PreSharedKey = cfg.psk;
+    }) // (optionalAttrs (cfg.passphrase != null) {
+      Security.Passprase = cfg.passphrase;
+    })));
+  });
 in {
-  options.networking.wireless.iwd.enable = mkEnableOption "iwd";
+  options.networking.wireless.iwd = {
+    enable = mkEnableOption "iwd";
+
+    interfaces = mkOption {
+      type = types.nullOr (types.listOf types.str);
+      example = [ "wlp2s0" ];
+      apply = x: if x == null then x else concatStringsSep " " x;
+      default = null;
+      description = ''
+        List of interfaces to use.
+      '';
+    };
+
+    networks = mkOption {
+      default = {};
+      example = {
+        "karlsruhe.freifunk.net" = {};
+        "secured".passphrase = "12345678";
+        "network-with-psk".psk = "bf4c086e055e492373a0f59ff614f61fc2d597a56db94e17a165f9ef1d42c066";
+      };
+
+      description = ''
+        Declarative configuration of wifi networks for
+        <citerefentry><refentrytitle>iwd</refentrytitle><manvolnum>8</manvolnum></citerefentry>.
+
+        All networks will be stored in
+        <literal>/var/lib/iwd/&lt;name&gt;.&lt;type&gt;</literal>.
+
+        Since each network is stored in its own file, declarative networks can be used in an
+        environment with imperatively added networks via
+        <citerefentry><refentrytitle>iwctl</refentrytitle><manvolnum>1</manvolnum></citerefentry>.
+      '';
+
+      type = types.attrsOf (types.submodule {
+        options = {
+          psk = mkOption {
+            type = types.nullOr (fixedStringLength 64 64);
+            default = null;
+            description = ''
+              WPA PSK for the currently defined network. This can be generated by using
+              <citerefentry><refentrytitle>wpa_passphrase</refentrytitle><manvolnum>8</manvolnum></citerefentry>.
+            '';
+          };
+
+          passphrase = mkOption {
+            type = types.nullOr (fixedStringLength 8 63);
+            default = null;
+            description = ''
+              Raw passphrase for the currently defined network.
+            '';
+          };
+
+          autoConnect = mkOption {
+            default = true;
+            type = types.bool;
+            apply = boolToString;
```

Comment:
`apply` should also be avoided here for the same reason (and similarly for the `hidden` option)